Stormcast customization conditions

jleinonen opened the pull request:
<!-- markdownlint-disable MD013-->
# PhysicsNeMo Pull Request

## Description
<!-- Provide a standalone description of changes in this PR. -->
<!-- Reference any issues closed by this PR with "closes #1234". -->
<!-- Note: The pull request title will be included in the CHANGELOG. -->

This PR adds support for configurable model conditions to the StormCast recipe.

**Main changes**:
1. Add configuration parameters `model.regression_conditions` and `model.diffusion_conditions`. These are lists consisting of `'state'`, `'background'`, `'invariant'` and/or `'regression'` (`'regression'` can only be specified in `model.diffusion_conditions`). The defaults are set to correspond to the StormCast paper.
2. Eliminate redundant parameters and configuration options that are now handled by the new two options: `previous_step_conditioning`, `use_regression_net`.
3. Streamline code to utilize the new parameters, eliminating redundancies. Particularly `utils.nn. build_network_condition_and_target` and `utils.nn.regression_model_forward`, as well as the calculation of model input dimension in `utils.trainer.py`, have been refactored.

**Features enabled by the changes**:
1. Training background-only or state-only models for benchmarking and debugging purposes, or to obtain a CorrDiff-like model to initialize a StormCast model state.
2. Other changes to model inputs, such as passing the background directly to the diffusion model.
3. Training pure diffusion (i.e. no regression step) models (if `'regression'` is not in `model.diffusion_conditions`).

## Checklist

- [x] I am familiar with the [Contributing Guidelines](https://github.com/NVIDIA/physicsnemo/blob/main/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.
- [ ] The [CHANGELOG.md](https://github.com/NVIDIA/physicsnemo/blob/main/CHANGELOG.md) is up to date with these changes.
- [ ] An [issue](https://github.com/NVIDIA/physicsnemo/issues) is linked to this pull request.

## Dependencies

<!-- Call out any new dependencies needed if any -->
No new dependencies are introduced.

pzharrington commented on examples/generative/stormcast/config/inference/stormcast.yaml:
```diff
@@ -15,7 +15,6 @@
 # limitations under the License.
 
 # General inference config items
-outdir: 'rundir' # Root path under which to save inference outputs
```
Was removing this intentional?

jleinonen replied:
This wasn't used anymore in the code (at least since the last PR, I'm not sure if it was used even before). All outputs are saved in `rundir` which I think is sufficient.

jleinonen replied:
Ah - I see, this is still needed in the config file even though it's not referred to in the code. I'll put it back in the next commit.

jleinonen replied:
Restored in https://github.com/NVIDIA/physicsnemo/pull/880/commits/4059d421f3028eebd628245e52f827e6d40aba9e.

pzharrington commented on examples/generative/stormcast/utils/nn.py:
```diff
@@ -65,27 +65,41 @@ def get_preconditioned_architecture(
+    regression_net=None,
+    condition_list=("state", "background"),
+    regression_condition_list=("state", "background"),
 ):
```
I think we should add a full docstring describing the behavior and inputs/outputs for this method since it is pretty important.

jleinonen replied:
Improved docstring and type hints added in https://github.com/NVIDIA/physicsnemo/pull/880/commits/0dd9c7e56868cbd11689edf5072b59c01e1107aa.

pzharrington reviewed (changes requested):
Overall the changes look nice and should improve usability/flexibility. I think the only needed additions are to flesh out the documentation a bit more:

- Let's add a complete docstring for `build_network_condition_and_target` since it is core to the training/inference functionality
- Document the new options and general setup/behavior in the README. Currently there are instructions for the old method that should be changed, and the descriptions of the model configs should also describe the/define the different components in the input condition list in this new approach

jleinonen commented:
> Overall the changes look nice and should improve usability/flexibility. I think the only needed additions are to flesh out the documentation a bit more:
> 
> * Let's add a complete docstring for `build_network_condition_and_target` since it is core to the training/inference functionality
> * Document the new options and general setup/behavior in the README. Currently there are instructions for the old method that should be changed, and the descriptions of the model configs should also describe the/define the different components in the input condition list in this new approach

I have now added a more complete docstring and described the new options in the README. I also added in the README more information about adding custom datasets and improved the docstring of the `StormCastDataset` class.

pzharrington reviewed (approved):
Nice work! Good to go

jleinonen commented:
/blossom-ci

jleinonen commented:
/blossom-ci
